src/misc/JsonConfig.cpp

```diff
@@ -0,0 +1,187 @@
+
+        // If failure to parse, we will create a new file
+        mIsNewConfig = true;
+        mJson = json::object();
+    }
+}
+
+void JsonConfig::SetArbitraryType(std::string key, json::value_type value) {
+    auto keyParts = StringHelper::Split(key, ".");
+
+    if (keyParts.size() > 1) {
+        // nested key
+        json* currentJson = &mJson;
+        for (size_t i = 0; i < keyParts.size() - 1; i++) {
+            if (!currentJson->contains(keyParts[i])) {
+                // create new object
+                (*currentJson)[keyParts[i]] = json::object();
+            }
+            currentJson = &(*currentJson)[keyParts[i]];
+        }
+        (*currentJson)[keyParts[keyParts.size() - 1]] = value;
+    } else {
+        // not nested
+        mJson[key] = value;
+    }
+}
+
+json::value_type JsonConfig::GetArbitraryType(std::string key) {
+    auto keyParts = StringHelper::Split(key, ".");
+
+    // find the deepest nested key if it exists
+    json* currentJson = &mJson;
+    for (size_t i = 0; i < keyParts.size() - 1; i++) {
+        if (!currentJson->contains(keyParts[i])) {
+            return nullptr;
+        }
+        currentJson = &(*currentJson)[keyParts[i]];
+    }
+
+    // extract the value if possible
+    if (currentJson->contains(keyParts[keyParts.size() - 1])) {
+        return (*currentJson)[keyParts[keyParts.size() - 1]];
+    } else {
+        return nullptr;
+    }
+}
+
+// MARK: - Public API
+
+bool JsonConfig::IsNewConfig() {
+    return mIsNewConfig;
+}
+
+void JsonConfig::DeleteEntry(std::string key) {
+    auto keyParts = StringHelper::Split(key, ".");
+
+    // clear nested key, if after deletion parent is empty, delete it too
+    if (keyParts.size() > 1) {
+        json* currentJson = &mJson;
+        for (size_t i = 0; i < keyParts.size() - 1; i++) {
+            if (!currentJson->contains(keyParts[i])) {
+                return;
+            }
+            currentJson = &(*currentJson)[keyParts[i]];
+        }
+        (*currentJson).erase(keyParts[keyParts.size() - 1]);
+
+        // delete empty parents
+        for (size_t i = keyParts.size() - 1; i > 0; i--) {
+            if ((*currentJson).empty()) {
+                currentJson = &mJson;
+                for (size_t j = 0; j < i - 1; j++) {
+                    currentJson = &(*currentJson)[keyParts[j]];
+                }
+                (*currentJson).erase(keyParts[i - 1]);
+            }
+        }
+    } else {
+        // not nested
+        mJson.erase(key);
+    }
+}
+
+json::value_type JsonConfig::GetRawEntry(std::string key) {
+    return GetArbitraryType(key);
+}
+
+void JsonConfig::SetInteger(std::string key, int32_t value) {
+    return SetArbitraryType(key, value);
+}
+
+int32_t JsonConfig::GetInteger(std::string key, int32_t defaultValue) {
+    json::value_type value = GetArbitraryType(key);
+    if (value.is_number_integer()) {
+        return value;
+    } else {
+        return defaultValue;
```

Comment:
i assume this is to handle the case where someone manually edited the json and put a value that doesn't parse to the correct type in there (no need to change anything, just hoping to confirm the reasoning behind these)